Cache/IO.lean

```diff
@@ -105,54 +109,39 @@ def isMathlibRoot : IO Bool :=
 
 section
 
-private def parseMathlibDepPath (json : Lean.Json) : Except String (Option FilePath) := do
-  let deps ← (← json.getObjVal? "packages").getArr?
-  for d in deps do
-    let n := ← (← d.getObjVal? "name").getStr?
-    if n != "mathlib" then
-      continue
-    let t := ← (← d.getObjVal? "type").getStr?
-    if t == "path" then
-      return some ⟨← (← d.getObjVal? "dir").getStr?⟩
-    else
-      return LAKEPACKAGESDIR / "mathlib"
-  return none
-
-private def CacheM.mathlibDepPath : IO FilePath := do
-  let raw ← IO.FS.readFile "lake-manifest.json"
-  match (Lean.Json.parse raw >>= parseMathlibDepPath) with
-  | .ok (some p) => return p
-  | .ok none =>
-      if ← isMathlibRoot then
-        return ⟨"."⟩
-      else
-        throw <| IO.userError s!"Mathlib not found in dependencies"
-  | .error e => throw <| IO.userError s!"Cannot parse lake-manifest.json: {e}"
+/-- Find path to `Mathlib` source directory -/
+private def CacheM.mathlibDepPath (sp : SearchPath) : IO FilePath := do
+  let mathlibSourceFile ← Lean.findLean sp `Mathlib
+  let some mathlibSource ← pure mathlibSourceFile.parent
+    | throw <| IO.userError s!"Mathlib not found in dependencies"
+  return mathlibSource
 
 -- TODO this should be generated automatically from the information in `lakefile.lean`.
 private def CacheM.getContext : IO CacheM.Context := do
-  let root ← CacheM.mathlibDepPath
-  return ⟨root, .ofList [
-    ("Mathlib", root),
-    ("Archive", root),
-    ("Counterexamples", root),
-    ("MathlibTest", root),
-    ("Aesop", LAKEPACKAGESDIR / "aesop"),
-    ("Batteries", LAKEPACKAGESDIR / "batteries"),
-    ("Cli", LAKEPACKAGESDIR / "Cli"),
-    ("ProofWidgets", LAKEPACKAGESDIR / "proofwidgets"),
-    ("Qq", LAKEPACKAGESDIR / "Qq"),
-    ("ImportGraph", LAKEPACKAGESDIR / "importGraph"),
-    ("LeanSearchClient", LAKEPACKAGESDIR / "LeanSearchClient"),
-    ("Plausible", LAKEPACKAGESDIR / "plausible")
-  ], LAKEPACKAGESDIR / "proofwidgets" / ".lake" / "build"⟩
+  let sp ← initSrcSearchPath
+  let mathlibSource ← CacheM.mathlibDepPath sp
+  return {
+    mathlibDepPath := mathlibSource,
+    srcSearchPath := sp,
+    packageDirs := .ofList [
+      ("Mathlib", mathlibSource),
```

Comment:
Should these all be parsed from the source path as well (in a future PR)?

Reply:
Indeed, #21238 will then use `let packageDir? := sp.find? (·.contains sourceFile)` instead of this `packageDirs`